### DISCUSSION TITLE
CAS, React and empty claims at first page load

### DISCUSSION THREAD

**Discussion by arnamq** [akunzai/GSS.Authentication.CAS#426] [Q&A] [answered]
Hi,

I've following Login method:

        [HttpGet("login")]
        public IActionResult Login(string returnUrl)
        {
            if (User.Identity?.IsAuthenticated == true)
            {
                return Redirect(returnUrl);
            }

            return Challenge(new AuthenticationProperties { RedirectUri = returnUrl }, "CAS");
        }

When I load the react.js page for a first time then after succesfull login to CAS User.Identity?.IsAuthenticated is always false. Next 3 requests to different  controller entpoints are causing another calls to login endpoint and finnaly some 500 errors. Sometimes lastcontroller entpoint call is succesful. It seems like controllers calls are made before first login call is finished and HttpContext.User.Identity isnt's yet set.

When I reload the page the Login endpoint isn't called at all and everything works fine.

I'm using GSS.Authentication.CAS.AspNetCore 5.3.3.

I setup CAS service like that:

    builder.Services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
        .AddCookie(options =>
        {
            options.Events.OnSigningOut = context =>
            {
                var redirectContext = new RedirectContext<CookieAuthenticationOptions>(
                    context.HttpContext,
                    context.Scheme,
                    context.Options,
                    context.Properties,
                    "/");

                if (builder.Configuration.GetValue("Authentication:CAS:SingleSignOut", false))
                {
                    // Single Sign-Out
                    var casUrl = new Uri(builder.Configuration["Authentication:CAS:ServerUrlBase"]);
                    var links = context.HttpContext.RequestServices.GetRequiredService<LinkGenerator>();
                    var serviceUrl = context.Properties.RedirectUri ?? links.GetUriByPage(context.HttpContext, "/");
                    redirectContext.RedirectUri = UriHelper.BuildAbsolute(
                        casUrl.Scheme,
                        new HostString(casUrl.Host, casUrl.Port),
                        casUrl.LocalPath, "/logout",
                        QueryString.Create("service", serviceUrl!));
                }

                context.Options.Events.RedirectToLogout(redirectContext);

                return Task.CompletedTask;
            };
            options.Cookie.Name = ".ASPXAUTH_EPROTOCOLS";
        })
        .AddCAS(options =>
        {
            options.CasServerUrlBase = builder.Configuration["Authentication:CAS:ServerUrlBase"];
            // required for CasSingleLogoutMiddleware
            options.SaveTokens = builder.Configuration.GetValue("Authentication:CAS:SaveTokens", false);
            var protocolVersion = builder.Configuration.GetValue("Authentication:CAS:ProtocolVersion", 3);

            if (protocolVersion != 3)
            {
                options.ServiceTicketValidator = protocolVersion switch
                {
                    1 => new Cas10ServiceTicketValidator(options),
                    2 => new Cas20ServiceTicketValidator(options),
                    _ => null
                };
            }

            options.Events.OnCreatingTicket = context =>
            {
                if (context.Identity == null)
                {
                    return Task.CompletedTask;
                }

                // Map claims from assertion
                var assertion = context.Assertion;

                context.Identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, assertion.PrincipalName));
                context.Identity.AddClaim(new Claim(ClaimTypes.Name, assertion.PrincipalName));

                if (assertion.Attributes.TryGetValue("display_name", out var displayName))
                {
                    context.Identity.AddClaim(new Claim(ClaimTypes.Name, displayName!));
                }

                if (assertion.Attributes.TryGetValue("email", out var email))
                {
                    context.Identity.AddClaim(new Claim(ClaimTypes.Email, email!));
                }

                return Task.CompletedTask;
            };
        });

Anyone can help?

**Comment by akunzai** [comment]
Have you check out the [ASP.NET Core sample app with React.js](https://github.com/akunzai/GSS.Authentication.CAS/blob/main/samples/AspNetCoreReactSample)?

It sounds like was the JavaScript asynchronous programming issue, are you using [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises)?

Please provide your errors and reproduce-able sample code.

**Comment by arnamq** [comment]
Yes, we're using Promise. 

Afer succesful login first request is OK:

![image](https://github.com/akunzai/GSS.Authentication.CAS/assets/26566399/2c74ea62-760e-4f30-90e4-455793ae4aaf)

but then next 3 API requests are causing another CAS calls:

![image](https://github.com/akunzai/GSS.Authentication.CAS/assets/26566399/b86f8002-1045-4a48-b270-2ab4310b7a99)

...and errors (http 500):

![image](https://github.com/akunzai/GSS.Authentication.CAS/assets/26566399/58397c19-09a9-4c67-ac57-41e1c91faf0c)

...but last API call is sucessful:

![image](https://github.com/akunzai/GSS.Authentication.CAS/assets/26566399/be447fd0-2ac0-4d16-bfb3-33ca35e5d25d)

**Comment by akunzai** [comment]
I couldn't figure out what the issue was based on your description and screenshots. Can you provide a minimal sample code?

**Comment by arnamq** [comment]
Which sample code do you need?  Our react.js code?

**Comment by akunzai** [comment]
Just like the [ASP.NET Core sample app with React.js](https://github.com/akunzai/GSS.Authentication.CAS/blob/main/samples/AspNetCoreReactSample), 

A runnable sample code can re-produce the issue!

**Comment by arnamq** [comment]
OK, we'll prepare some sample app and let you know.

**Comment by arnamq** [comment]
Hi @akunzai, 

We've prepared some demo app but also noticed that both production and demo apps works fine if runned as standalone kestrel apps. If we run them on IIS then we have issue as mentioned above.

Our demo app: [eProtocols.Simplified.zip](https://github.com/akunzai/GSS.Authentication.CAS/files/12857232/eProtocols.Simplified.zip)

**Comment by akunzai** [comment, marked as answer]
If this only happens on IIS,

perhaps you should check web.config inheritance problems.

- https://stackoverflow.com/q/782252
- https://learn.microsoft.com/aspnet/core/host-and-deploy/iis/web-config#configuration-of-aspnet-core-module-with-webconfig

**Comment by arnamq** [comment]
Hi @akunzai,

You Was right - It was some kind of inheritance problem. Thank you for your support!
